elm/benchmark_diff.py

```diff
@@ -0,0 +1,161 @@
+from benchmarks import parity_reference, eval_code_string
+from elm.util.diff_eval import split_diff, apply_diff
+
+
+parity_test_data = {
+            i: parity_reference(*i) for i in itertools.product(range(2), repeat=4)
+        }
+
+
+def eval_completions(
+    codes: Union[str, list[str]], task: str = "parity", timeout: int = 5
+) -> Union[int, list[int]]:
+    """
+    Evaluate (a batch of) the modified codes on a task.
+    Args:
+        codes: either a string or a batch of strings. The code(s) to be evaluated.
+        task: (Optional) the task to be performed.
+        timeout: (Optional) the timeout (in seconds).
+    Returns:
+        either the status code of the string or a list of status codes of the batch
+        of strings (depending on whether `codes` is batched).
+    """
+    if task == "parity":
+        _codes = codes if isinstance(codes, list) else [codes]
+        results = []
+        for code in _codes:
+            results.append(eval_code_string(code, parity_test_data, timeout))
+        if isinstance(codes, list):
+            # Batch evaluation returns the batch
+            return results
+        else:
+            # Single evaluation returns a single result
+            return results[0]
+    else:
+        raise ValueError(f"Unknown task: {task}")
+
+
+def mutate_code(n_bugs: int = 5, task: str = "parity") -> tuple:
+    """
+    Mutate code to create n bugs. Output the prompt in diff format.
+    Args:
+        n_bugs: number of bugs to introduce (from 1 to 5).
+        task: (Optional) the task to be performed.
+    Returns:
+        mutated_code, function_string
+    """
+    mutation_template = [
+        f"<NME> {task}.py\n<BEF> ",
+        "\n<MSG> Fixed bugs",
+    ]
+    if task == "parity":
+        variables = ["b", "b", "b", "b", 2]
+        for i in range(n_bugs):
+            variables[i] = "c" if i < 4 else 3
+        func_str = (
+            'def parity(b1,b2,b3,b4):\n    """Return binary parity of a sequence of input bits.'
+            ' Return 0 for even parity, 1 for odd parity."""\n    bit_sum = sum(['
+            "{}1,{}2,{}3,{}4])\n    return bit_sum % {}".format(*variables)
+        )
+        mutation_template.insert(1, func_str)
+        return "".join(mutation_template), func_str
+    else:
+        raise ValueError(f"Unknown task: {task}")
+
+
+@hydra.main(
+    config_path=str(SRC_PATH / "config"),
+    config_name="benchmark_diff_cfg",
+    version_base="1.2",
+)
+def main(cfg):
+    print("----------------- Config ---------------")
+    print(OmegaConf.to_yaml(cfg))
+    print("-----------------  End -----------------")
+
+    device = torch.device("cuda" if cfg.cuda else "cpu")
+    config = AutoConfig.from_pretrained(cfg.model)
+    # Sometimes our model just fresh came out of training. Force use_cache to be true.
+    config.use_cache = True
+    tokenizer = AutoTokenizer.from_pretrained(cfg.model)
+    if cfg.gpus > 1:
+        model = torch.nn.DataParallel(
```

Comment:
Oh I want to make a note here that DataParallel doesn't work (even for 1 node). This applies to codegen util https://github.com/CarperAI/OpenELM/blob/281429b8b9f9ca955239c1ec617eb62bced280f5/elm/codegen/codegen_utilities.py#L111
as well. Better way (also recommended by pytorch) is to use DDP but that would make the code a little complicated.